crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -181,56 +227,781 @@ impl<S: States> NextState<S> {
+            }
+        }
+    }
+}
+
+fn execute_state_transitions(world: &mut World) {
+    prepare_state_transitions(world);
+    let _ = world.try_run_schedule(ManualStateTransitions);
+    while let Some((_, compute)) = world
+        .get_resource_mut::<StateTransitionSchedules>()
+        .and_then(|mut schedules| schedules.dependant_schedules.pop_first())
+    {
+        for schedule in compute {
+            let _ = world.try_run_schedule(schedule);
+        }
+    }
+    execute_state_transition_schedules(world);
+}
+
+#[derive(Resource)]
+struct StateTransitionsHaveBeenSetup;
+
+/// Sets up the schedules and systems for handling state transitions
+/// within a [`World`].
+///
+/// Runs automatically when using `App` to insert states, but needs to
+/// be added manually in other situations.
+pub fn setup_state_transitions_in_world(world: &mut World) {
+    if world
+        .get_resource::<StateTransitionsHaveBeenSetup>()
+        .is_some()
+    {
+        return;
+    }
+    world.insert_resource(StateTransitionsHaveBeenSetup);
+    let mut schedules = world.get_resource_or_insert_with(Schedules::default);
+    let mut schedule = Schedule::new(StateTransition);
+    schedule.add_systems(execute_state_transitions);
+    schedules.insert(schedule);
 }
 
 /// If a new state is queued in [`NextState<S>`], this system:
 /// - Takes the new state value from [`NextState<S>`] and updates [`State<S>`].
 /// - Sends a relevant [`StateTransitionEvent`]
 /// - Runs the [`OnExit(exited_state)`] schedule, if it exists.
 /// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
+/// - Derive any dependant states through the [`ComputeDependantStates::<S>`] schedule, if it exists.
 /// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
-pub fn apply_state_transition<S: States>(world: &mut World) {
+pub fn apply_state_transition<S: FreelyMutableState>(world: &mut World) {
     // We want to take the `NextState` resource,
     // but only mark it as changed if it wasn't empty.
-    let Some(mut next_state_resource) = world.get_resource_mut::<NextState<S>>() else {
+    let Some(next_state_resource) = world.get_resource::<NextState<S>>() else {
         return;
     };
-    if let Some(entered) = next_state_resource.bypass_change_detection().0.take() {
-        next_state_resource.set_changed();
-        match world.get_resource_mut::<State<S>>() {
-            Some(mut state_resource) => {
-                if *state_resource != entered {
-                    let exited = mem::replace(&mut state_resource.0, entered.clone());
-                    world.send_event(StateTransitionEvent {
-                        before: exited.clone(),
-                        after: entered.clone(),
+
+    match next_state_resource {
+        NextState::Set(new_state) => {
+            let new_state = new_state.clone();
+            internal_apply_state_transition(world, Some(new_state));
+        }
+        _ => {
```

Comment:
Nit: we should avoid using a blanket match here: it's not any faster and makes future refactoring harder.

Reply:
done